ci/terraform/waf.tf

```diff
@@ -83,114 +83,205 @@ resource "aws_wafv2_web_acl" "frontend_cloudfront_waf_web_acl" {
-      block {}
-    }
+  dynamic "rule" {
+    for_each = local.ip_endpoint_rate_limiting_configuration
+    content {
+      name     = "BlockMoreThan${rule.value.limit}RequestsFromIpPer${rule.value.evaluation_window_sec}Seconds"
+      priority = 20 + rule.key
+      rule_label {
+        name = "TooManyRequestsFromIpOnSpecificEndpoints"
+      }
 
-    statement {
-      rate_based_statement {
-        limit                 = var.environment == "staging" ? 20000000 : var.rate_limited_endpoints_requests_per_period
-        evaluation_window_sec = var.rate_limited_endpoints_rate_limit_period
-        aggregate_key_type    = "IP"
+      action {
+        block {}
+      }
 
+      statement {
+        rate_based_statement {
+          limit                 = rule.value.limit
+          evaluation_window_sec = rule.value.evaluation_window_sec
+          aggregate_key_type    = "CUSTOM_KEYS"
+          custom_key {
+            ip {}
+          }
+          custom_key {
+            uri_path {
+              text_transformation {
+                priority = 0
+                type     = "URL_DECODE"
+              }
+              text_transformation {
+                priority = 1
+                type     = "LOWERCASE"
+              }
+            }
+          }
 
-        scope_down_statement {
-          or_statement {
-            dynamic "statement" {
-              for_each = var.rate_limited_endpoints
+          scope_down_statement {
+            dynamic "or_statement" {
+              // If there are more than 1 provided endpoints, use an or_statement
+              for_each = length(rule.value.endpoints) > 1 ? [1] : []
               content {
-                byte_match_statement {
-                  positional_constraint = "STARTS_WITH"
-                  search_string         = statement.value
-                  field_to_match {
-                    uri_path {}
-                  }
-                  text_transformation {
-                    priority = 0
-                    type     = "LOWERCASE"
+                dynamic "statement" {
+                  for_each = rule.value.endpoints
+                  content {
+                    byte_match_statement {
+                      positional_constraint = "STARTS_WITH"
+                      search_string         = lower(statement.value)
+                      field_to_match {
+                        uri_path {}
+                      }
+                      text_transformation {
+                        priority = 0
+                        type     = "URL_DECODE"
+                      }
+                      text_transformation {
+                        priority = 1
+                        type     = "LOWERCASE"
+                      }
+                    }
                   }
                 }
               }
             }
+            dynamic "byte_match_statement" {
+              // If there's only 1 endpoint, just create the single byte_match_statement
```

Comment:
I presume this duplication is because `or_statement` doesn't support just 1 `byte_match_statement`

Reply:
Correct. Previously we had to add a second 'fake' endpoint to protect for the or to work, but now it'll gracefully handle a single entry without that hack